library/include/rocwmma/internal/b32_util.hpp

```diff
@@ -0,0 +1,161 @@
+
+            static_assert(SelectIdx < (sizeof(ExtractedT) / sizeof(EmplacedT)),
+                          "Invalid index selection");
+
+            for(uint32_t i = 0u; i < decltype(rIt)::range(); i++, rIt++, wIt++)
+            {
+                FootSpace a;
+                a.emplaced[SelectIdx] = get<0>(*rIt);
+                get<0>(*wIt)          = a.extracted;
+            }
+            return result;
+        }
+
+        template <uint32_t VecSize, uint32_t SelectIdx = 0u>
+        __device__ static auto emplace(VecT<ExtractedT, VecSize> const& v)
+        {
+            VecT<EmplacedT, VecSize> result;
+            auto const               rIt = makeVectorIterator(v).begin();
+            auto                     wIt = makeVectorIterator(result).begin();
+
+            static_assert(decltype(rIt)::range() == decltype(wIt)::range(),
+                          "Unexpected iterator range mismatch");
+
+            static_assert(SelectIdx < (sizeof(ExtractedT) / sizeof(EmplacedT)),
+                          "Invalid index selection");
+
+            for(uint32_t i = 0u; i < decltype(rIt)::range(); i++, rIt++, wIt++)
+            {
+                FootSpace a;
+                a.extracted  = get<0>(*rIt);
+                get<0>(*wIt) = a.emplaced[SelectIdx];
+            }
+            return result;
+        }
+    };
+
+    template <>
+    struct b32Util<float64_t>
+    {
+        using ExtractedT = float32_t;
+        using EmplacedT  = float64_t;
+
+        using FootSpace = union
+        {
+            ExtractedT extracted[2];
+            EmplacedT  emplaced;
+        };
+
+        template <uint32_t VecSize, uint32_t SelectIdx = 0u>
+        __device__ static auto extract(VecT<EmplacedT, VecSize> const& v)
+        {
+            const uint32_t inflatedSize = VecSize << 1;
+
+            VecT<ExtractedT, inflatedSize> result;
+            auto const                     rIt = makeVectorIterator(v).begin();
+            auto                           wIt = makeVectorIterator(result).begin();
+
+            static_assert(decltype(rIt)::range() == (decltype(wIt)::range() >> 1),
+                          "Unexpected iterator range mismatch");
+
+            static_assert(SelectIdx < (sizeof(EmplacedT) / sizeof(ExtractedT)),
+                          "Invalid index selection");
+
+            for(uint32_t i = 0u; i < decltype(rIt)::range(); i++, rIt++, wIt++)
+            {
+                FootSpace a;
+                a.emplaced         = get<0>(*rIt);
+                get<0>(*wIt)       = a.extracted[0];
+                get<VecSize>(*wIt) = a.extracted[1];
+            }
+            return result;
+        }
+
+        template <uint32_t VecSize, uint32_t SelectIdx = 0u>
+        __device__ static auto emplace(VecT<ExtractedT, VecSize> const& v)
+        {
+            const uint32_t inflatedSize = VecSize << 1;
+
+            VecT<EmplacedT, VecSize> result;
```

Comment:
Just like above, the inflated size won't change, we just need to choose whether the final result will emplace the input into upper or lower half of f64.